src/ingest/src/main/scala/osmesa/ingest/ProcessOSM.scala

```diff
@@ -304,6 +308,141 @@ object ProcessOSM {
+
+            dissolveRings(retryRings ++ remaining, dissolvedOuters ++ newRings.map(_.exterior), dissolvedInners ++ newRings.flatMap(_.holes))
+        }
+    }
+  }
+
+  private def dissolveRings(rings: List[Line]): (List[Line], List[Line]) = {
+    dissolveRings(rings.map(x => Polygon(x)))
+  }
+
+  val buildMultiPolygon: UserDefinedFunction = udf((id: Long, ways: Seq[Row]) => {
+    try {
+      val coords: Seq[(String, Line)] = ways
+        .map(row => (row.getAs[String]("role"), row.getAs[Array[Byte]]("geom").readWKB match {
+          case geom: Polygon => geom.as[Polygon].map(_.exterior)
+          case geom => geom.as[Line]
+        })).filter(_._2.isDefined).map(x => (x._1, x._2.get))
+
+      val (completeOuters, completeInners, partialOuters, partialInners, completeUnknowns, partialUnknowns) = coords.foldLeft((List.empty[Line], List.empty[Line], List.empty[Line], List.empty[Line], List.empty[Line], List.empty[Line])) {
+        case ((co, ci, po, pi, cu, pu), (role, geom: Line)) =>
+          Option(geom) match {
+            case Some(line) =>
+              role match {
+                case "outer" if line.isClosed => (co :+ line, ci, po, pi, cu, pu)
+                case "outer" => (co, ci, po :+ line, pi, cu, pu)
+                case "inner" if line.isClosed => (co, ci :+ line, po, pi, cu, pu)
+                case "inner" => (co, ci, po, pi :+ line, cu, pu)
+                case "" if line.isClosed => (co, ci, po, pi, cu :+ line, pu)
+                case "" => (co, ci, po, pi, cu, pu :+ line)
+                case _ => (co, ci, po, pi, cu, pu)
+              }
+            case None => (co, ci, po, pi, cu, pu)
+          }
+      }
+
+      val unknowns: List[Line] = completeUnknowns ++ connectSegments(partialUnknowns)
+
+      val (outers, inners) = unknowns.foldLeft((completeOuters ++ connectSegments(partialOuters), completeInners ++ connectSegments(partialInners))) {
+        case ((o: List[Line], i: List[Line]), u) =>
+          if (o.exists(Polygon(_).contains(u))) {
+            (o, i :+ u)
+          } else {
+            (o :+ u, i)
+          }
+      }
+
+      // reclassify rings according to their topology (ignoring roles)
+      val (classifiedOuters, classifiedInners) = (outers ++ inners).map(Polygon(_)).sortWith(_.area > _.area) match {
+        case h :: t => t.foldLeft((List(h), List.empty[Polygon])) {
+          case ((os, is), ring) =>
+            ring match {
+              // there's an outer ring that contains this one; this is an inner ring
+              case _ if os.exists(or => or.contains(ring)) => (os, is :+ ring)
+              case _ => (os :+ ring, is)
+            }
+        }
+        case Nil => (List.empty[Polygon], List.empty[Polygon])
+      }
+
+      val (dissolvedOuters, addlInners) = dissolveRings(classifiedOuters)
+      val (dissolvedInners, _) = dissolveRings(classifiedInners.map(_.exterior) ++ addlInners)
```

Comment:
What should be done with holes in resulting multipolygons? Should they be treated as additional inner rings?